<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SpatialExperiment with tidy principles • tidySpatialExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="SpatialExperiment with tidy principles">
<meta property="og:description" content="tidySpatialExperiment provides a bridge between the SpatialExperiment package and the tidyverse ecosystem. It creates an invisible layer that allows you to interact with a SpatialExperiment object as if it were a tibble; enabling the use of functions from dplyr, tidyr, ggplot2 and plotly. But, underneath, your data remains a SpatialExperiment object.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidySpatialExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/overview.html">Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/william-hutchison/tidySpatialExperiment/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="tidyspatialexperiment---part-of-tidyomics-">tidySpatialExperiment - part of <em>tidyomics</em> <img src="reference/figures/logo.png" id="tidySpatialExperiment_logo" align="right" width="125"><a class="anchor" aria-label="anchor" href="#tidyspatialexperiment---part-of-tidyomics-"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>tidySpatialExperiment provides a bridge between the <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a> package and the <a href="https://www.tidyverse.org" class="external-link"><em>tidyverse</em></a> ecosystem. It creates an invisible layer that allows you to interact with a <code>SpatialExperiment</code> object as if it were a tibble; enabling the use of functions from <a href="https://github.com/tidyverse/dplyr" class="external-link">dplyr</a>, <a href="https://github.com/tidyverse/tidyr" class="external-link">tidyr</a>, <a href="https://github.com/tidyverse/ggplot2" class="external-link">ggplot2</a> and <a href="https://github.com/plotly/plotly.R" class="external-link">plotly</a>. But, underneath, your data remains a <code>SpatialExperiment</code> object.</p>
<p>tidySpatialExperiment also provides five additional utility functions.</p>
<div class="section level2">
<h2 id="resources">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h2>
<p>If you would like to learn more about tidySpatialExperiment and <em>tidyomics</em>, the following links are a good place to start:</p>
<ul>
<li><a href="http://william-hutchison.github.io/tidySpatialExperiment/" class="external-link">The tidySpatialExperiment website</a></li>
<li><a href="https://github.com/tidyomics" class="external-link">The tidyomics website</a></li>
</ul>
<p>The <em>tidyomics</em> ecosystem also includes packages for:</p>
<ul>
<li>
<p>Working with genomic features:</p>
<ul>
<li>
<a href="https://github.com/sa-lee/plyranges" class="external-link">plyranges</a>, for tidy manipulation of genomic range data.</li>
<li>
<a href="https://github.com/nullranges/nullranges" class="external-link">nullranges</a>, for tidy generation of genomic ranges representing the null hypothesis.</li>
<li>
<a href="https://github.com/tidyomics/plyinteractions" class="external-link">plyinteractions</a>, for tidy manipulation of genomic interaction data.</li>
</ul>
</li>
<li>
<p>Working with transcriptomic features:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a>, for tidy manipulation of <code>SummarizedExperiment</code> objects.</li>
<li>
<a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a>, for tidy manipulation of <code>SingleCellExperiment</code> objects.</li>
<li>
<a href="https://github.com/stemangiola/tidyseurat" class="external-link">tidyseurat</a>, for tidy manipulation of <code>Seurat</code> objects.</li>
<li>
<a href="https://github.com/stemangiola/tidybulk" class="external-link">tidybulk</a>, for bulk RNA-seq analysis.</li>
</ul>
</li>
<li>
<p>Working with cytometry features:</p>
<ul>
<li>
<a href="https://github.com/keyes-timothy/tidytof" class="external-link">tidytof</a>, for tidy manipulation of high-dimensional cytometry data.</li>
</ul>
</li>
<li>
<p>And a few associated packages:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidygate" class="external-link">tidygate</a>, for manual gating of points in space.</li>
<li>
<a href="https://github.com/stemangiola/tidyHeatmap/" class="external-link">tidyheatmap</a>, for modular heatmap contruction.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="functions-and-utilities">Functions and utilities<a class="anchor" aria-label="anchor" href="#functions-and-utilities"></a>
</h2>
<table class="table">
<colgroup>
<col width="9%">
<col width="90%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Functions available</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>SpatialExperiment</code></td>
<td>All</td>
</tr>
<tr class="even">
<td><code>dplyr</code></td>
<td>
<code>arrange</code>,<code>bind_rows</code>, <code>bind_cols</code>, <code>distinct</code>, <code>filter</code>, <code>group_by</code>, <code>summarise</code>, <code>select</code>, <code>mutate</code>, <code>rename</code>, <code>left_join</code>, <code>right_join</code>, <code>inner_join</code>, <code>slice</code>, <code>sample_n</code>, <code>sample_frac</code>, <code>count</code>, <code>add_count</code>
</td>
</tr>
<tr class="odd">
<td><code>tidyr</code></td>
<td>
<code>nest</code>, <code>unnest</code>, <code>unite</code>, <code>separate</code>, <code>extract</code>, <code>pivot_longer</code>
</td>
</tr>
<tr class="even">
<td><code>ggplot2</code></td>
<td><code>ggplot</code></td>
</tr>
<tr class="odd">
<td><code>plotly</code></td>
<td><code>plot_ly</code></td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="20%">
<col width="79%">
</colgroup>
<thead><tr class="header">
<th>Utility</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>as_tibble</code></td>
<td>Convert cell data to a <code>tbl_df</code>
</td>
</tr>
<tr class="even">
<td><code>join_features</code></td>
<td>Append feature data to cell data</td>
</tr>
<tr class="odd">
<td><code>aggregate_cells</code></td>
<td>Aggregate cell-feature abundance into a pseudobulk <code>SummarizedExperiment</code> object</td>
</tr>
<tr class="even">
<td><code>rectangle</code></td>
<td>Select cells in a rectangular region of space</td>
</tr>
<tr class="odd">
<td><code>ellipse</code></td>
<td>Select cells in an elliptical region of space</td>
</tr>
<tr class="even">
<td><code>gate_spatial</code></td>
<td></td>
</tr>
<tr class="odd">
<td><code>gate_programmatic</code></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the stable version of tidySpatialExperiment from Bioconductor.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidySpatialExperiment"</span><span class="op">)</span></span></code></pre></div>
<p>Or, you can install the development version of tidySpatialExperiment from GitHub.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"pak"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"pak"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pak.html" class="external-link">pak</a></span><span class="op">(</span><span class="st">"william-hutchison/tidySpatialExperiment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>Here, we attach tidySpatialExperiment and an example <code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example SpatialExperiment object</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/william-hutchison/tidySpatialExperiment" class="external-link">tidySpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">read10xVisium</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="spatialexperiment-tibble-abstraction">SpatialExperiment-tibble abstraction<a class="anchor" aria-label="anchor" href="#spatialexperiment-tibble-abstraction"></a>
</h2>
<p>A <code>SpatialExperiment</code> object represents assay-feature values as rows and cells as columns. Additional information about the cells is stored in the <code>reducedDims</code>, <code>colData</code> and <code>spatialCoords</code> slots.</p>
<p>tidySpatialExperiment provides a SpatialExperiment-tibble abstraction, representing cells as rows and cell data as columns, in accordance with the tidy observation-variable convention. The cell data is made up of information stored in the <code>colData</code> and <code>spatialCoords</code> slots.</p>
<p>The default view is now of the SpatialExperiment-tibble abstraction.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 7</span></span>
<span><span class="co">#  # Features = 50 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>But, our data maintains its status as a <code>SpatialExperiment</code> object. Therefore, we have access to all <code>SpatialExperiment</code> functions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">#                     in_tissue array_row array_col   sample_id</span></span>
<span><span class="co">#                     &lt;logical&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">#  AAACAACGAATAGTTC-1     FALSE         0        16    section1</span></span>
<span><span class="co">#  AAACAAGTATCTCCCA-1      TRUE        50       102    section1</span></span>
<span><span class="co">#  AAACAATCTACTAGCA-1      TRUE         3        43    section1</span></span>
<span><span class="co">#  AAACACCAATAACTGC-1      TRUE        59        19    section1</span></span>
<span><span class="co">#  AAACAGAGCGACTCCT-1      TRUE        14        94    section1</span></span>
<span><span class="co">#  AAACAGCTTTCAGAAG-1     FALSE        43         9    section1</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#                     pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#  AAACAACGAATAGTTC-1               2312               1252</span></span>
<span><span class="co">#  AAACAAGTATCTCCCA-1               8230               7237</span></span>
<span><span class="co">#  AAACAATCTACTAGCA-1               4170               1611</span></span>
<span><span class="co">#  AAACACCAATAACTGC-1               2519               8315</span></span>
<span><span class="co">#  AAACAGAGCGACTCCT-1               7679               2927</span></span>
<span><span class="co">#  AAACAGCTTTCAGAAG-1               1831               6400</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  DataFrame with 1 row and 4 columns</span></span>
<span><span class="co">#      sample_id    image_id   data scaleFactor</span></span>
<span><span class="co">#    &lt;character&gt; &lt;character&gt; &lt;list&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#  1    section1      lowres   ####   0.0510334</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="integration-with-the-tidyverse-ecosystem">Integration with the <em>tidyverse</em> ecosystem<a class="anchor" aria-label="anchor" href="#integration-with-the-tidyverse-ecosystem"></a>
</h1>
<div class="section level2">
<h2 id="manipulate-with-dplyr">Manipulate with dplyr<a class="anchor" aria-label="anchor" href="#manipulate-with-dplyr"></a>
</h2>
<p>Most functions from dplyr are available for use with the SpatialExperiment-tibble abstraction. For example, <code><a href="reference/filter.html">filter()</a></code> can be used to filter cells by a variable of interest.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>   <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">array_col</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 3 × 7</span></span>
<span><span class="co">#  # Features = 50 | Cells = 3 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACATGGTGAGAGGA-1 FALSE            62         0 section1                1212</span></span>
<span><span class="co">#  2 AAACGAAGATGGAGTA-1 FALSE            58         4 section1                1487</span></span>
<span><span class="co">#  3 AAAGAATGACCTTAGA-1 FALSE            64         2 section1                1349</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>And <code>mutate</code> can be used to add new variables, or modify the value of an existing variable.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>in_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">&amp;</span> <span class="va">array_row</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 8</span></span>
<span><span class="co">#  # Features = 50 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell     in_tissue array_row array_col sample_id in_region pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;     &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;lgl&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACG… FALSE             0        16 section1  FALSE                   2312</span></span>
<span><span class="co">#  2 AAACAAGT… TRUE             50       102 section1  FALSE                   8230</span></span>
<span><span class="co">#  3 AAACAATC… TRUE              3        43 section1  TRUE                    4170</span></span>
<span><span class="co">#  4 AAACACCA… TRUE             59        19 section1  FALSE                   2519</span></span>
<span><span class="co">#  5 AAACAGAG… TRUE             14        94 section1  FALSE                   7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tidy-with-tidyr">Tidy with tidyr<a class="anchor" aria-label="anchor" href="#tidy-with-tidyr"></a>
</h2>
<p>Most functions from tidyr are also available. Here, <code><a href="reference/nest.html">nest()</a></code> is used to group the data by <code>sample_id</code>, and <code><a href="reference/unnest.html">unnest()</a></code> is used to ungroup the data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Nest the SpatialExperiment object by sample_id</span></span>
<span><span class="va">spe_nested</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the nested SpatialExperiment object</span></span>
<span><span class="va">spe_nested</span></span>
<span><span class="co">#  # A tibble: 1 × 2</span></span>
<span><span class="co">#    sample_id data           </span></span>
<span><span class="co">#    &lt;chr&gt;     &lt;list&gt;         </span></span>
<span><span class="co">#  1 section1  &lt;SptlExpr[,50]&gt;</span></span>
<span></span>
<span><span class="co"># Unnest the nested SpatialExperiment objects</span></span>
<span><span class="va">spe_nested</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 7</span></span>
<span><span class="co">#  # Features = 50 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plot-with-ggplot2">Plot with ggplot2<a class="anchor" aria-label="anchor" href="#plot-with-ggplot2"></a>
</h2>
<p>The <code><a href="reference/ggplot.html">ggplot()</a></code> function can be used to create a plot directly from a <code>SpatialExperiment</code> object. This example also demonstrates how tidy operations can be combined to build up more complex analysis.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span> <span class="op">&amp;</span> <span class="va">in_tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Add a column with the sum of feature counts per cell</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>count_sum <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_int</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="op">~</span></span>
<span>        <span class="va">spe</span><span class="op">[</span>, <span class="va">.x</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu">counts</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span></span>
<span>      <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Plot with tidySpatialExperiment and ggplot2</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">count_sum</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">count_sum</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-11-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="plot-with-plotly">Plot with plotly<a class="anchor" aria-label="anchor" href="#plot-with-plotly"></a>
</h2>
<p>The <code><a href="reference/plot_ly.html">plot_ly()</a></code> function can also be used to create a plot from a <code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="op">~</span> <span class="va">array_col</span>, </span>
<span>        y <span class="op">=</span> <span class="op">~</span> <span class="va">array_row</span>, </span>
<span>        color <span class="op">=</span> <span class="op">~</span> <span class="va">in_tissue</span>, </span>
<span>        type <span class="op">=</span> <span class="st">"scatter"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/plotly_demo.png"></p>
</div>
</div>
<div class="section level1">
<h1 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a>
</h1>
<div class="section level2">
<h2 id="append-feature-data-to-cell-data">Append feature data to cell data<a class="anchor" aria-label="anchor" href="#append-feature-data-to-cell-data"></a>
</h2>
<p>The <em>tidyomics</em> ecosystem places an emphasis on interacting with cell data. To interact with feature data, the <code><a href="reference/join_features.html">join_features()</a></code> function can be used to append assay-feature values to cell data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Join feature data in wide format, preserving the SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000025915"</span>, <span class="st">"ENSMUSG00000042501"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 9</span></span>
<span><span class="co">#  # Features = 6 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id ENSMUSG00000025915</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;dbl&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                   0</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                   0</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                   0</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                   0</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                   0</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 3 more variables: ENSMUSG00000042501 &lt;dbl&gt;, pxl_col_in_fullres &lt;int&gt;,</span></span>
<span><span class="co">#  #   pxl_row_in_fullres &lt;int&gt;</span></span>
<span></span>
<span><span class="co"># Join feature data in long format, discarding the SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000025915"</span>, <span class="st">"ENSMUSG00000042501"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"long"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  tidySpatialExperiment says: A data frame is returned for independent data </span></span>
<span><span class="co">#                analysis.</span></span>
<span><span class="co">#  # A tibble: 6 × 7</span></span>
<span><span class="co">#    .cell       in_tissue array_row array_col sample_id .feature .abundance_counts</span></span>
<span><span class="co">#    &lt;chr&gt;       &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;                &lt;dbl&gt;</span></span>
<span><span class="co">#  1 AAACAACGAA… FALSE             0        16 section1  ENSMUSG…                 0</span></span>
<span><span class="co">#  2 AAACAACGAA… FALSE             0        16 section1  ENSMUSG…                 0</span></span>
<span><span class="co">#  3 AAACAAGTAT… TRUE             50       102 section1  ENSMUSG…                 0</span></span>
<span><span class="co">#  4 AAACAAGTAT… TRUE             50       102 section1  ENSMUSG…                 1</span></span>
<span><span class="co">#  5 AAACAATCTA… TRUE              3        43 section1  ENSMUSG…                 0</span></span>
<span><span class="co">#  # ℹ 1 more row</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="aggregate-cells">Aggregate cells<a class="anchor" aria-label="anchor" href="#aggregate-cells"></a>
</h2>
<p>Sometimes, it is necessary to aggregate the gene-transcript abundance from a group of cells into a single value. For example, when comparing groups of cells across different samples with fixed-effect models.</p>
<p>The <code><a href="reference/aggregate_cells.html">aggregate_cells()</a></code> function can be used to aggregate cells by a specified variable and assay, returning a <code>SummarizedExperiment</code> object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/aggregate_cells.html">aggregate_cells</a></span><span class="op">(</span><span class="va">in_tissue</span>, assays <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="co">#  class: SummarizedExperiment </span></span>
<span><span class="co">#  dim: 50 2 </span></span>
<span><span class="co">#  metadata(0):</span></span>
<span><span class="co">#  assays(1): counts</span></span>
<span><span class="co">#  rownames(50): ENSMUSG00000002459 ENSMUSG00000005886 ...</span></span>
<span><span class="co">#    ENSMUSG00000104217 ENSMUSG00000104328</span></span>
<span><span class="co">#  rowData names(1): feature</span></span>
<span><span class="co">#  colnames(2): FALSE TRUE</span></span>
<span><span class="co">#  colData names(3): in_tissue .aggregated_cells sample_id</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="elliptical-and-rectangular-region-selection">Elliptical and rectangular region selection<a class="anchor" aria-label="anchor" href="#elliptical-and-rectangular-region-selection"></a>
</h2>
<p>The <code><a href="reference/ellipse.html">ellipse()</a></code> and <code><a href="reference/rectangle.html">rectangle()</a></code> functions can be used to select cells by their position in space.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>in_ellipse <span class="op">=</span> <span class="fu"><a href="reference/ellipse.html">ellipse</a></span><span class="op">(</span><span class="va">array_col</span>, <span class="va">array_row</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">40</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_col</span>, y <span class="op">=</span> <span class="va">array_row</span>, colour <span class="op">=</span> <span class="va">in_ellipse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-15-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="interactive-gating">Interactive gating<a class="anchor" aria-label="anchor" href="#interactive-gating"></a>
</h2>
<p>For the interactive selection of cells in space, tidySpatialExperiment experiment provides <code><a href="reference/gate.html">gate()</a></code>. This function uses <a href="https://github.com/stemangiola/tidygate" class="external-link">tidygate</a>, shiny and plotly to launch an interactive plot overlaying cells in position with image data. Additional parameters can be used to specify point colour, shape, size and alpha, either with a column in the SpatialExperiment object or a constant value.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_gated</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/gate.html">gate</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"in_tissue"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/gate_interactive_demo.gif"></p>
<p>A record of which points appear in which gates is appended to the SpatialExperiment object in the <code>.gated</code> column. To select cells which appear within any gates, filter for non-NA values. To select cells which appear within a specific gate, string pattern matching can be used.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select cells within any gate</span></span>
<span><span class="va">spe_gated</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.gated</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 4 × 8</span></span>
<span><span class="co">#  # Features = 50 | Cells = 4 | Assays = counts</span></span>
<span><span class="co">#    .cell        in_tissue array_row array_col sample_id .gated pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;        &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;               &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACGAGACGG… TRUE             35        79 section1  2                    6647</span></span>
<span><span class="co">#  2 AAACTGCTGGC… TRUE             45        67 section1  2                    5821</span></span>
<span><span class="co">#  3 AAAGGGATGTA… TRUE             24        62 section1  1,2                  5477</span></span>
<span><span class="co">#  4 AAAGGGCAGCT… TRUE             24        26 section1  1                    3000</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span>
<span><span class="co"># Select cells within gate 2</span></span>
<span><span class="va">spe_gated</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">.gated</span>, <span class="st">"2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 3 × 8</span></span>
<span><span class="co">#  # Features = 50 | Cells = 3 | Assays = counts</span></span>
<span><span class="co">#    .cell        in_tissue array_row array_col sample_id .gated pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;        &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;               &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACGAGACGG… TRUE             35        79 section1  2                    6647</span></span>
<span><span class="co">#  2 AAACTGCTGGC… TRUE             45        67 section1  2                    5821</span></span>
<span><span class="co">#  3 AAAGGGATGTA… TRUE             24        62 section1  1,2                  5477</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>Details of the interactively drawn gates are saved to <code>tidygate_env$gates</code>. This variable is overwritten each time interactive gates are drawn, so save it right away if you would like to access it later.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect previously drawn gates</span></span>
<span><span class="va">tidygate_env</span><span class="op">$</span><span class="va">gates</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  # A tibble: 6 × 3</span></span>
<span><span class="co">#        x     y .gate</span></span>
<span><span class="co">#    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#  1 4310. 3125.     1</span></span>
<span><span class="co">#  2 3734. 3161.     1</span></span>
<span><span class="co">#  3 2942. 3521.     1</span></span>
<span><span class="co">#  4 2834. 3665.     1</span></span>
<span><span class="co">#  5 2834. 4385.     1</span></span>
<span><span class="co">#  # ℹ 1 more row</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save if needed</span></span>
<span><span class="va">tidygate_env</span><span class="op">$</span><span class="va">gates</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">write_rds</span><span class="op">(</span><span class="st">"important_gates.rds"</span><span class="op">)</span></span></code></pre></div>
<p>If previously drawn gates are supplied to the <code>programmatic_gates</code> argument, cells will be gated programmatically. This feature allows the reproduction of previously drawn interactive gates.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>important_gates <span class="ot">&lt;-</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>  <span class="fu">read_rds</span>(<span class="st">"important_gates.rds"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>spe <span class="sc">|&gt;</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>  <span class="fu">gate</span>(<span class="at">programmatic_gates =</span> important_gates)<span class="er">)</span> <span class="sc">|&gt;</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(.gated))</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">#  # A SpatialExperiment-tibble abstraction: 4 × 8</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="co">#  # Features = 50 | Cells = 4 | Assays = counts</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="co">#    .cell        in_tissue array_row array_col sample_id .gated pxl_col_in_fullres</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co">#    &lt;chr&gt;        &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;               &lt;int&gt;</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="co">#  1 AAACGAGACGG… TRUE             35        79 section1  2                    6647</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="co">#  2 AAACTGCTGGC… TRUE             45        67 section1  2                    5821</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co">#  3 AAAGGGATGTA… TRUE             24        62 section1  1,2                  5477</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="co">#  4 AAAGGGCAGCT… TRUE             24        26 section1  1                    3000</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="special-column-behaviour">Special column behaviour<a class="anchor" aria-label="anchor" href="#special-column-behaviour"></a>
</h1>
<p>Removing the <code>.cell</code> column will return a tibble. This is consistent with the behaviour in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">.cell</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  tidySpatialExperiment says: Key columns are missing. A data frame is </span></span>
<span><span class="co">#                          returned for independent data analysis.</span></span>
<span><span class="co">#  # A tibble: 6 × 4</span></span>
<span><span class="co">#    in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#    &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;    </span></span>
<span><span class="co">#  1 FALSE             0        16 section1 </span></span>
<span><span class="co">#  2 TRUE             50       102 section1 </span></span>
<span><span class="co">#  3 TRUE              3        43 section1 </span></span>
<span><span class="co">#  4 TRUE             59        19 section1 </span></span>
<span><span class="co">#  5 TRUE             14        94 section1 </span></span>
<span><span class="co">#  # ℹ 1 more row</span></span></code></pre></div>
<p>The <code>sample_id</code> column cannot be removed with <em>tidyverse</em> functions, and can only be modified if the changes are accepted by SpatialExperiment’s <code>colData()</code> function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample_id is not removed, despite the user's request</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 7</span></span>
<span><span class="co">#  # Features = 50 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span>
<span></span>
<span><span class="co"># This change maintains separation of sample_ids and is permitted</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="st">"_modified"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 7</span></span>
<span><span class="co">#  # Features = 6 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id  pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                   &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1_…               2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1_…               8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1_…               4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1_…               2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1_…               7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span>
<span></span>
<span><span class="co"># This change does not maintain separation of sample_ids and produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="st">"new_sample"</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 50 × 7</span></span>
<span><span class="co">#  # Features = 50 | Cells = 50 | Assays = counts</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id  pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                   &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 new_sample               2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 new_sample               8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 new_sample               4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 new_sample               2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 new_sample               7679</span></span>
<span><span class="co">#  # ℹ 45 more rows</span></span>
<span><span class="co">#  # ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>The <code>pxl_col_in_fullres</code> and <code>px_row_in_fullres</code> columns cannot be removed or modified with <em>tidyverse</em> functions. This is consistent with the behaviour of dimension reduction data in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attempting to remove pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span>
<span><span class="co">#  Error in `select_helper()`:</span></span>
<span><span class="co">#  ! Can't select columns that don't exist.</span></span>
<span><span class="co">#  ✖ Column `pxl_col_in_fullres` doesn't exist.</span></span>
<span></span>
<span><span class="co"># Attempting to modify pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span>
<span><span class="co">#  Error in `dplyr::mutate()`:</span></span>
<span><span class="co">#  ℹ In argument: `pxl_col_in_fullres`.</span></span>
<span><span class="co">#  Caused by error:</span></span>
<span><span class="co">#  ! object 'pxl_col_in_fullres' not found</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<p>If you use tidySpatialExperiment in published research, please cite <a href="https://doi.org/10.1038/s41592-024-02299-2" class="external-link">The tidyomics ecosystem: enhancing omic data analyses</a>.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/tidySpatialExperiment" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidySpatialExperiment</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>William Hutchison <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0001-6242-4269" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Stefano Mangiola <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-blue.svg" alt="Lifecycle:experimental"></a></li>
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/actions" class="external-link"><img src="https://github.com/william-hutchison/tidySpatialExperiment/workflows/rworkflows/badge.svg" alt="R build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by William Hutchison, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
